File: Codes/Matlab/Chapters/Chapter6/SIR_Tornado.m

% Created on Mon Jan  3 17:42:39 2022
% Model of SIR
% @author: Cogan

clear;
close all
set(groot,'DefaultAxesTickLabelInterpreter','latex');
set(groot,'DefaultLegendInterpreter','latex');



p_names=[ 'k','gamma'];
params=[ 0.005, 1];
N=1000;
tspan = linspace(0, 10, 500);

[t,y_solution]=ode45(@(t,Y) rhs(t,Y,params), [tspan(1),tspan(end)], [N,1,0]); 
plot(t, y_solution(:,1),'k','LineWidth',2)
hold on
plot(t, y_solution(:,2),'k','LineWidth',2)
plot(t, y_solution(:,3),'k','LineWidth',2)

legend(['Susceptible','Infected','Recovered'], fontsize=16)
xlabel('Time', fontsize=16)
ylabel('Population', fontsize=16)

figure(2)
params1=[ 0.005,2];

[t,y_solution]=ode45(@(t,Y) rhs(t,Y,params), [tspan(1),tspan(end)], [N,1,0]); 
plot(t, y_solution(:,1),'k','LineWidth',2)
hold on
plot(t, y_solution(:,2),'b','LineWidth',2)
plot(t, y_solution(:,3),'c','LineWidth',2)

legend(['Susceptible','Infected','Recovered'], fontsize=16)
xlabel('Time', fontsize=16)
ylabel('Population', fontsize=16)


%Generate alpha plot
figure(3)
params1=[ 0.01, 0.5];
[t2,y_solution2]=ode45(@(t,Y) rhs(t,Y,params1), [tspan(1),tspan(end)], [.75*N,1,0]); 
[t3,y_solution3]=ode45(@(t,Y) rhs(t,Y,params1), [tspan(1),tspan(end)], [.5*N,1,0]); 
[t4,y_solution4]=ode45(@(t,Y) rhs(t,Y,params1), [tspan(1),tspan(end)], [.25*N,1,0]); 
[t5,y_solution5]=ode45(@(t,Y) rhs(t,Y,params1), [tspan(1),tspan(end)], [.05*N,1,0]); 

plot(t2, y_solution2(:,2),'k','LineWidth',2)
hold on
plot(t3, y_solution3(:,2),'b','LineWidth',2)
plot(t4, y_solution4(:,2),'c','LineWidth',2)
plot(t5, y_solution5(:,2),'m','LineWidth',2)
legend('$\alpha = .75$','$\alpha = .5$','$\alpha = .25$','$\alpha = .05$', fontsize=16)
xlabel('Time', fontsize=16)
ylabel('Population', fontsize=16)


%===============================
%      Tornado plot
% we need low and high values of 
% parameters with results Qoi_low 
% and QoI_high
%===============================




params_t=[.1, 2000, .01, .01, .5, 1000-10,10,0]; %augment parameters for initial conditions
params_min_t=[.1, 2000, .01, .01, .5, 1000-10,10,0];
params_max_t=[.1, 2000, .01, .01, .5, 1000-10,10,0];
QoI_min=zeros(1,7);
QoI_max=zeros(1,7);

for k =1:7
    params_min_t(k)=.5*params_min_t(k);
    params_max_t(k)=1.5*params_max_t(k);
    [t2_min_t,y_solution_min_t]=ode45(@(t,Y) rhs_t(t,Y,params_min_t), [tspan(1),tspan(end)], [params_t(end-2),params_t(end-1), params_t(end)]); 
    [t2_max_t,y_solution_max_t]=ode45(@(t,Y) rhs_t(t,Y,params_max_t), [tspan(1),tspan(end)], [params_t(end-2),params_t(end-1), params_t(end)]); 
    QoI_min(k)=y_solution_min_t(end,1)/(y_solution_min_t(end,1)+y_solution_min_t(end,2));
    QoI_max(k)=y_solution_max_t(end,1)/(y_solution_max_t(end,1)+y_solution_max_t(end,2));
    params_t=[.1, 2000, .01, .01, .5, 1000-10,10,0]; %augment parameters for initial conditions
    params_min_t=[.1, 2000, .01, .01, .5, 1000-10,10,0];
    params_max_t=[.1, 2000, .01, .01, .5, 1000-10,10,0];
end

%========================
%Make the tornado plots
%========================
[QoI_min,ind]=sort(QoI_min,'descend');
QoI_max=QoI_max(ind);

figure
h = barh(QoI_min);
hold on
barh(-QoI_max,'r')

xticklabels({[[max(QoI_max),max(QoI_max)/2, 0,max(QoI_max)/2, max(QoI_max)]]})
yticklabels({'$r$','$\kappa$','$\gamma$','$\delta$','$S_0$','$I_0$','$R_0$'})

%=========================
%Define the RHS of ODEs
%=========================
function f1 = rhs_S(t,Y,params)
    S=Y(1);
    I=Y(2);
    R=Y(3);
    k=params(1);
    gamma=params(2);
    f1 = -k*S*I;
end

function f2 = rhs_I(t,Y,params)
    S=Y(1);
    I=Y(2);
    R=Y(3);
    k=params(1);
    gamma=params(2);
    f2 = k*S*I-gamma*I;
end

function f3 = rhs_R(t,Y,params)
    S=Y(1);
    I=Y(2);
    R=Y(3);
    k=params(1);
    gamma=params(2);
    f3 = gamma*I;
end
function g = rhs(t,Y,params)
    S=Y(1);
    I=Y(2);
    R=Y(3);
    k=params(1);
    gamma=params(2);
    g = [rhs_S(t,Y,params);rhs_I(t,Y,params);rhs_R(t,Y,params)];
end

%=========================
%Define the RHS of ODEs
%for tornado plots
%=========================
function f1 = rhs_S_t(t,Y,params)
    S=Y(1);
    I=Y(2);
    R=Y(3);
    r=params(1);
    kappa=params(2);
    k=params(3);
    gamma=params(4);
    delta=params(5);
    IC_S=params(6);
    IC_I=params(7);
    IC_R=params(8);
    f1 = -k*S*I;
end

function f2 = rhs_I_t(t,Y,params)
    S=Y(1);
    I=Y(2);
    R=Y(3);
    r=params(1);
    kappa=params(2);
    k=params(3);
    gamma=params(4);
    delta=params(5);
    IC_S=params(6);
    IC_I=params(7);
    IC_R=params(8);
    f2 = k*S*I-gamma*I;
end

function f3 = rhs_R_t(t,Y,params)
    S=Y(1);
    I=Y(2);
    R=Y(3);
    r=params(1);
    kappa=params(2);
    k=params(3);
    gamma=params(4);
    delta=params(5);
    IC_S=params(6);
    IC_I=params(7);
    IC_R=params(8);
    f3 = gamma*I;
end
function g = rhs_t(t,Y,params)
    S=Y(1);
    I=Y(2);
    R=Y(3);
    r=params(1);
    kappa=params(2);
    k=params(3);
    gamma=params(4);
    delta=params(5);
    IC_S=params(6);
    IC_I=params(7);
    IC_R=params(8);
    g = [rhs_S_t(t,Y,params);rhs_I_t(t,Y,params);rhs_R_t(t,Y,params)];
end

File: Codes/Python/Chapters/Chapter6/SIR_Tornado.py
"""
Created on Mon Jan  3 17:42:39 2022
Model of SIR
@author: Cogan
"""


from IPython import get_ipython
get_ipython().magic('reset -sf')
import numpy as np
import matplotlib.pyplot as pyplot
from scipy.integrate import odeint
from scipy.optimize import fsolve
from scipy.integrate import solve_ivp
import random
pyplot.close('all')
pyplot.style.use('tableau-colorblind10')  #Fixes colorscheme to be accessible



#Define the RHS of ODEs
def f1(t,Y,params):
    S,I,R=Y
    k, gamma = params
    return -k*S*I
def f2(t,Y,params):
    S,I,R=Y
    k, gamma = params
    return k*S*I-gamma*I
def f3(t,Y,params):
    S,I,R=Y
    k, gamma = params
    return gamma*I
def rhs(t,Y,params):
    S,I,R=Y
    k, gamma = params
    return [f1(t,Y,params),f2(t,Y,params),f3(t,Y,params)]
# same as usual: define f and g, rhs returns the right hand side
# parameters are from the paper, beware the order though
p_names=( 'k','gamma')
params=np.array([ 0.005, 1])
N=1000
tspan = np.linspace(0, 10, 500)
yp= solve_ivp(lambda t,Y: rhs(t,Y,params), [tspan[0],tspan[-1]], [N,1,0], method='RK45',t_eval=tspan)
pyplot.figure(1)

pyplot.plot(tspan,yp.y[0])
pyplot.plot(tspan,yp.y[1])
pyplot.plot(tspan,yp.y[2])
pyplot.legend(['Susceptible','Infected','Recovered'], fontsize=16)
pyplot.legend(['Susceptible','Infected','Recovered'], fontsize=16)
pyplot.xlabel('Time', fontsize=16)
pyplot.ylabel('Population', fontsize=16)
#pyplot.savefig('SIR_high.png')

pyplot.figure(2)
params1=np.array([ 0.005,2])
yp1= solve_ivp(lambda t,Y: rhs(t,Y,params1), [tspan[0],tspan[-1]], [N,1,0], method='LSODA',t_eval=tspan)
pyplot.plot(tspan,yp1.y[0])
pyplot.plot(tspan,yp1.y[1])
pyplot.plot(tspan,yp.y[2])
pyplot.legend(['Susceptible','Infected','Recovered'], fontsize=16)
pyplot.xlabel('Time', fontsize=16)
pyplot.ylabel('Population', fontsize=16)
#pyplot.savefig('SIR_low.png')


###GENERATE alpha plot
pyplot.figure(3)
params1=np.array([ 0.01, 0.5])
yp2= solve_ivp(lambda t,Y: rhs(t,Y,params1), [tspan[0],tspan[-1]], [.75*N,1,0], method='RK45',t_eval=tspan)
yp3= solve_ivp(lambda t,Y: rhs(t,Y,params1), [tspan[0],tspan[-1]], [.5*N,1,0], method='RK45',t_eval=tspan)
yp4= solve_ivp(lambda t,Y: rhs(t,Y,params1), [tspan[0],tspan[-1]], [.25*N,1,0], method='RK45',t_eval=tspan)
yp5= solve_ivp(lambda t,Y: rhs(t,Y,params1), [tspan[0],tspan[-1]], [.05*N,1,0], method='RK45',t_eval=tspan)

pyplot.plot(tspan,yp2.y[1])
pyplot.plot(tspan,yp3.y[1])
pyplot.plot(tspan,yp4.y[1])
pyplot.plot(tspan,yp5.y[1])

pyplot.legend([r'$\alpha = .75$',r'$\alpha = .5$',r'$\alpha = .25$',r'$\alpha = .05$'], fontsize=16)
pyplot.xlabel('Time', fontsize=16)
pyplot.ylabel('Population', fontsize=16)
#pyplot.savefig('SIR_alpha.png')


###############################################################################
#Tornado plot
#we need low and high values of parameters with results Qoi_low and QoI_high
###############################################################################




#Define the RHS of ODEs for the tornado plots. We have expanded the number of parameters
def f1_t(t,Y,params):
    S,I,R=Y
    r, kappa, k, gamma, delta, IC_S, IC_I, IC_R = params
    return r*S*(kappa-S)-k*S*I
def f2_t(t,Y,params):
    S,I,R=Y
    r, kappa, k, gamma, delta , IC_S, IC_I, IC_R= params
    return k*S*I-gamma*I-delta*I
def f3_t(t,Y,params):
    S,I,R=Y
    r, kappa, k, gamma, delta , IC_S, IC_I, IC_R= params
    return gamma*I
def rhs_t(t,Y,params):
    S,I,R=Y
    r, kappa, k, gamma, delta , IC_S, IC_I, IC_R= params
    return [f1_t(t,Y,params),f2_t(t,Y,params),f3_t(t,Y,params)]


tspan = np.linspace(0, 50, 1000)

params_t=[.1, 2000, .01, .01, .5, 1000-10,10,0] #augment parameters for initial conditions
params_min_t=[.1, 2000, .01, .01, .5, 1000-10,10,0]
params_max_t=[.1, 2000, .01, .01, .5, 1000-10,10,0]
QoI_min=np.zeros([7])
QoI_max=np.zeros([7])

for k in np.arange(0,7):
    params_min_t[k]=.5*params_min_t[k]
    params_max_t[k]=1.5*params_max_t[k]
    y_solution_min_t = solve_ivp(lambda t,Y: rhs_t(t,Y,params_min_t),  [tspan[0],tspan[-1]],[params_t[-3],params_t[-2], params_t[-1]],method='RK45',t_eval=tspan)
    y_solution_max_t = solve_ivp(lambda t,Y: rhs_t(t,Y,params_max_t),  [tspan[0],tspan[-1]],[params_t[-3],params_t[-2], params_t[-1]],method='RK45',t_eval=tspan)     
    QoI_min[k]=y_solution_min_t.y[0][-1]/(y_solution_min_t.y[0][-1]+y_solution_min_t.y[1][-1])
    QoI_max[k]=y_solution_max_t.y[0][-1]/(y_solution_max_t.y[0][-1]+y_solution_max_t.y[1][-1])
    params_t=[.1, 2000, .01, .01, .5, 1000-10,10,0] #augment parameters for initial conditions
    params_min_t=[.1, 2000, .01, .01, .5, 1000-10,10,0]
    params_max_t=[.1, 2000, .01, .01, .5, 1000-10,10,0]


#Make the tornado plots
pos_label = (['$r$','$\kappa$','$\gamma$','$\delta$','$S_0$','$I_0$','$R_0$'])#
pos=np.arange(7) + .5

fig, (ax_left, ax_right) = pyplot.subplots(ncols=2)
ax_left.barh(pos, QoI_min, align='center')
ax_left.set_yticks([])
ax_left.invert_xaxis()
ax_right.barh(pos, QoI_max, align='center')
ax_right.set_yticks(pos)
ax_right.set_yticklabels(pos_label, ha='center', x=-1)
ax_right.set_xlim((0,.0075))
ax_left.set_xlim((0.0075,0))
pyplot.savefig('SIR_tornado.png')























